src/libAtomVM/mailbox.h

```diff
@@ -256,20 +269,22 @@ Message *mailbox_first(Mailbox *mbox);
 /**
  * @brief Free memory associated with a mailbox.
  *
- * @details All messages in the mailbox will be freed.
+ * @details All messages in the mailbox will be freed or appended to the heap.
  * @param mbox the mailbox to free.
+ * @param heap the heap to add messages to.
  */
-void mailbox_destroy(Mailbox *mbox);
+void mailbox_destroy(Mailbox *mbox, Heap *heap);
 
 /**
- * @brief Free memory associated with a signal mailbox message.
+ * @brief Dispose a (processed) mailbox message. The message will be freed or
+ * appended to current heap and will be destroyed on garbage collect.
+ * This function is called by mailbox_remove and is only needed for signal
+ * messages.
  *
- * @details The supplied message will be free'd, and
- * to any references to shared memory will decrement
- * reference counts.
  * @param m the message to free.
+ * @param heap heap to append the message to.
  */
-void mailbox_destroy_signal_message(MailboxMessage *m);
+void mailbox_message_dispose(MailboxMessage *m, Heap *heap);
```

Comment:
ok, I think this interesting, I should be able to close  #467 (it should work in a similar way to `mailbox_take_first`).
So I should be able to:
- dispose the message, so I get a heap fragment
- pass the heap fragment to another thread
- process it
- destroy the heap fragment from the other thread?

Do you think it can work?
Is it possible to free the heap fragment without running any GC pass?

Reply:
The current pattern for ports indeed is:

```
Message *message = mailbox_first(&ctx->mailbox);
...
mailbox_remove_message(&ctx->mailbox, &ctx->heap);
```

And `mailbox_remove_message` removes the message from the list and then calls `mailbox_message_dispose`.

So we could allow ports to get the first message, remove it from the list and then be responsible for calling disposing the message.

If they take any term from the message and put it in a heap, they will need to make the message a heap fragment and merge it into the target heap (i.e. call `mailbox_message_dispose`) to ensure the terms they took are safe and garbage collectable.

If they don't pass any term from the message and don't do anything with heaps, no GC pass is necessary and they could simply free the message.